src/pyhf/parameters/paramview.py

```diff
@@ -1,114 +1,107 @@
 from .. import get_backend, default_backend, events
+from ..tensor.common import _TensorViewer
 
 
-def index_helper(name, tensor_shape, batch_shape, par_map):
-    if isinstance(name, list):
-        return [index_helper(x, tensor_shape, batch_shape, par_map) for x in name]
-    x = list(range(int(default_backend.product(tensor_shape))))
-    indices = default_backend.reshape(x, tensor_shape)
-    parfield_slice = tuple(slice(None, None) for x in batch_shape) + (
-        par_map[name]['slice'],
+def _tensorviewer_from_parmap(par_map, batch_size):
+    db = default_backend
+    # prepares names and per-parset ranges
+    # in the order or the parameters
+    names, indices, starts = list(
+        zip(
+            *sorted(
+                [
+                    (
+                        k,
+                        db.astensor(range(v['slice'].start, v['slice'].stop)),
+                        v['slice'].start,
+                    )
+                    for k, v in par_map.items()
+                ],
+                key=lambda x: x[2],
+            )
+        )
     )
-    indices = indices[parfield_slice]
-    return default_backend.tolist(indices)
+    return _TensorViewer(indices, names=names, batch_size=batch_size)
+
+
+def _tensorviewer_from_slices(slices, names, batch_size):
+    target_slices = []
+    start = 0
+    for sl in slices:
+        stop = start + (sl.stop - sl.start)
+        target_slices.append(slice(start, stop))
+        start = stop
+
+    db = default_backend
+    ranges = []
+    for sl in target_slices:
+        ranges.append(db.astensor(range(sl.start, sl.stop)))
+    if not ranges:
+        return None
+    return _TensorViewer(ranges, names=names, batch_size=batch_size)
+
+
+def extract_index_access(
```

Comment:
what does this do?

Reply:
it prepares index tensors for `tb.gather(parameters,indextensor)`-like operations